Scripts/SOD: Implement Sylvanas Windrunner's introduction

Nyr97 opened the pull request:
**A brief note about worldstates**: I'm not exactly sure if Testing State is the correct naming as it's impossible to know. However, it might refer as well as to Closed State instead, being 0 false and 1 true. This would lead us to believe that Open State could be related instead to Release Week reached or not, in which 1 is true and 0 false. That would explain why Normal and Heroic difficulties share the same worldstate, as well as Mythic and the first dungeon of Looking for Raid since they get released at the same time in pairs. 

In any case, this is just my research and thoughts, you might as well think differently, feel free to share your opinion in the comments.

**Changes proposed:**

- Added instance script for Sanctum of Domination.
- Added worldstates to allow entrance to the different difficulty modes.
- Added initial spawns for creatures present during the introduction and their texts during the encounter.
- Added initial spawns for gameobjects present during the introduction.
- Created and added several serverside areatriggers to cause death on players when falling down the different platforms of the encounter.
- Implemented Sylvanas Windrunner's Introduction conversation.
- Added localization esES support for warning messages during the encounter.

**Issues addressed:**

None.

**Tests performed:**

It builds and it was tested in-game.

**Known issues and TODO list:**

- [x] Find out if her encounter ends with her npcId or a spell on players, most likely a spell since she's feigning death. In any case, I wasn't able to find anything on the sniff. Could it be serverside?
- [x] Find out why even after setting the correct worldstates, it won't allow entrance without GM on.
- [x] Find out whether the teleport from the creature Throne of the Damned goes back to Kel'thuzad's room or the raid's entrance.
- [x] Sylvanas Windrunner's creature_texts for Black Arrow (Mythic only) are missing.

Nyr97 committed:
Scripts/SOD: Implement Sylvanas Windrunner's introduction

* added instanceScript for Sanctum of Domination.
* added initial spawns for creatures and their texts and gameobjects in Phase 1.
* added serverside areatrigger to cause death on players when falling down the different platforms of the encounter.

Nyr97 committed:
missing info on comments

Nyr97 committed:
build fix

Nyr97 committed:
fix nopch build

Nyr97 committed:
minor additions

* added world states for LFR.
* added lastEncounterDungeon for Sylvanas Windrunner.

Nyr97 committed:
codestyle

Nyr97 committed:
added introduction trigger

Nyr97 committed:
Implemented Throne of the Damned

Nyr97 committed:
fixed a bug where Throne of the Damned was not clickable

Nyr97 committed:
fixed a bug where portal wouldn't show up correctly after a reset

mdX7 commented on src/server/scripts/Shadowlands/SanctumOfDomination/boss_sylvanas_windrunner.cpp:
```diff
@@ -0,0 +1,443 @@
+    }
+
+private:
+    InstanceScript* _instance;
```
Unused

Nyr97 replied:
Feedback applied!

mdX7 commented on src/server/scripts/Shadowlands/SanctumOfDomination/boss_sylvanas_windrunner.cpp:
```diff
@@ -0,0 +1,443 @@
+
+    void OnUnitEnter(Unit* unit) override
+    {
+        if (!_instance || !unit->IsPlayer() || (_instance->GetData(DATA_SYLVANAS_INTRO) == IN_PROGRESS || _instance->GetData(DATA_SYLVANAS_INTRO) == DONE))
```
Should be faster if you are checking if `_instance->GetData(DATA_SYLVANAS_INTRO)` is != the required state since its only triggered once anyway right?

Nyr97 replied:
I assume that would be better? In that case, I should be using `_instance->GetData(DATA_SYLVANAS_INTRO) == NOT_STARTED`?

Nyr97 replied:
Feedback applied!

mdX7 commented on src/server/scripts/Shadowlands/SanctumOfDomination/sanctum_of_domination.h:
```diff
@@ -0,0 +1,112 @@
+
+#include "CreatureAIImpl.h"
+
+#define DataHeader "SD"
```
SOD

Nyr97 replied:
Feedback applied!

Nyr97 committed:
Feedback applied

Nyr97 committed:
Black Arrow texts for Sylvanas

Nyr97 committed:
removed instance check from Z check areatrigger

Nyr97 committed:
spelling

Nyr97 committed:
space for sql

Nyr97 committed:
removed hardcoded value from SQL

Nyr97 committed:
sql standard fix

Nyr97 committed:
added missing texts for warnings and localization support for esES

Nyr97 committed:
removed wrong data parsed from sql

Nyr97 committed:
adjustments

* fixed a bug where Throne of the Damned couldn't be found since it was not declared in creatureData.
* fixed a bug where Introduction would be started every time the areatrigger was triggered because it failed to obtain IN_PROGRESS data.

Nyr97 committed:
used Status for SylvanasIntroductionData

Nyr97 committed:
Throne of the Damned is now controlled via SetBossState

Nyr97 committed:
removed unused enum

Nyr97 committed:
Throne of the Damned now works correctly

Nyr97 committed:
removed unused param

mdX7 commented on src/server/scripts/Shadowlands/SanctumOfDomination/boss_sylvanas_windrunner.cpp:
```diff
@@ -0,0 +1,429 @@
+                    break;
+
+                case EVENT_INTRODUCTION + 3:
+                    me->SendPlaySpellVisualKit(SPELL_VISUAL_KIT_SYLVANAS_INTRODUCTION_TALK_02, 0, 0);
```
i assume this should be executed when she reached POINT_INTRO_01?

Nyr97 replied:
It should happen mid-way iirc.

mdX7 committed:
prepare merge 1

mdX7 commented:
ref encounter end credit, related quest id is 63903 which requires criteriatreeid 91083 fulfilled

Nyr97 committed:
remove note

Nyr97 committed:
Scripts/SOD: added LoadDungeonEncounterData

* handles credit upon completion of encounters.

Nyr97 committed:
misspellings

Nyr97 committed:
oopsies

Nyr97 commented:
merge when

mdX7 commented:
Very soontm

Nyr97 committed:
Merge remote-tracking branch 'upstream/master' into SOD-Sylvanas-Introduction

Nyr97 committed:
oopsies

Nyr97 committed:
update worldstates

Nyr97 committed:
update worldstates again

Nyr97 committed:
Merge remote-tracking branch 'upstream/master' into SOD-Sylvanas-Introduction

Nyr97 committed:
Scripts/SoD: few changes

* Sylvanas' introduction conversation and events were moved into the conversation's update method entirely to reduce the main boss' AI size.
* changed raw setting of unitflags for helpers methods.
* removed some notes.
* added creature_template_difficulty entries for LFR, normal and heroic.
NOTE: Death Gate is not spawned in LFR because the last hall is only one boss and it doesn't allow players moving away from the boss' areas (this doesn't apply to the Death Gate after the encounter is done, which allows the player to move back to the Sanctum of Domination's entrance).

Nyr97 committed:
Scripts/SOD: added EventMap start event

Nyr97 committed:
Scripts/SoD: lining enums

Nyr97 committed:
fix nopch

Nyr97 committed:
fix nopch

Nyr97 committed:
oopsies

Nyr97 committed:
oopsies

Nyr97 committed:
SQL/SOD: updates to difficulty templates

* UNKNOWN fields are yet to be sniffed and parsed.

Nyr97 committed:
SQL/SoD: changed UNKNOWN fields to 99999999

Nyr97 committed:
Merge remote-tracking branch 'upstream/master' into SOD-Sylvanas-Introduction

Nyr97 committed:
SQL/SoD: updates

* updated creature_template_difficulty with Mythic entries.
* added some creature_template that were parsed.
* added two new texts for sylvanas.

Nyr97 committed:
Scripts/SoD: clean-up

Nyr97 committed:
Update 9999_99_99_99_world.sql

Nyr97 committed:
Update 9999_99_99_99_world.sql

Nyr97 committed:
Merge remote-tracking branch 'upstream/master' into SOD-Sylvanas-Introduction

Nyr97 committed:
finished creature_template_difficulty

mdX7 committed:
Merge remote-tracking branch 'remotes/origin/master' into teleqraph/SOD-Sylvanas-Introduction

mdX7 committed:
prepare merge

mdX7 committed:
prepare merge2

mdX7 committed:
prepare merge3
